File: public/dsa/searching/matrix_search/optimal_2/matrix_search.cpp
#include <bits/stdc++.h>
using namespace std;

bool searchMatrix(vector<vector<int>> &matrix, int key) {
    int rows = matrix.size();
    int cols = matrix[0].size();

    int l = 0;
    int h = rows * cols - 1;

    while (l <= h) {
        int m = (l + h) / 2;
        int v = matrix[m / cols][m % cols];

        if (v == key) {
            return true;
        } else if (v < key) {
            l = m + 1;
        } else {
            h = m - 1;
        }
    }
    return false;
}

// tests-start
int main() {
    vector<vector<int>> matrix1 = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9},
    };

    vector<vector<int>> matrix2 = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9},
        {10, 11, 12},
    };

    cout << boolalpha << searchMatrix(matrix1, 6) << endl;
    cout << boolalpha << searchMatrix(matrix2, 15) << endl;

    return 0;
}

// tests-end

/*output
true
false
*/

File: public/dsa/searching/matrix_search/optimal_2/matrix_search.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], key: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])

        l = 0
        h = rows * cols - 1

        while l <= h:
            m = (l + h) // 2
            v = matrix[m // cols][m % cols]

            if v == key:
                return True
            elif v < key:
                l = m + 1
            else:
                h = m - 1

        return False


# tests-start
matrix1 = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
]

matrix2 = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [10, 11, 12],
]

sol = Solution()
print(sol.searchMatrix(matrix1, 6))
print(sol.searchMatrix(matrix2, 15))

# tests-end


"""output
True
False
"""
